src/librustc_mir/borrow_check/diagnostics/mutability_errors.rs

```diff
@@ -257,12 +257,22 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
                 assert_eq!(local_decl.mutability, Mutability::Not);
 
                 err.span_label(span, format!("cannot {ACT}", ACT = act));
-                err.span_suggestion(
-                    local_decl.source_info.span,
-                    "consider changing this to be mutable",
-                    format!("mut {}", self.local_names[local].unwrap()),
-                    Applicability::MachineApplicable,
-                );
+
+                if let ty::Ref(_, _, Mutability::Mut) = local_decl.ty.kind {
```

Comment:
This condition seems insufficient. You'd need to check that the expected type requires `&mut T` and not `&mut &mut T`, otherwise it seems like it would trigger on this (which is a good test to add, to ensure that this behaves correctly):
```rust
struct S;

fn bar(x: &mut &mut S) {}

fn foo() {
    let x = &mut S;
    bar(&mut x);
}
```

...but this makes me wonder whether this is doable in borrowck if you don't know what type is expected (seems you only know that you want to take a mutable borrow and cannot due to lack of `mut` on the binding. On the other hand, if you do check it in typeck, then we need to check that:
```rust
struct S;

fn bar(x: &mut S) {}

fn foo() {
    let mut x = &mut S;
    bar(&mut x);
}
```
still compiles... which is probably not a good idea in typeck.